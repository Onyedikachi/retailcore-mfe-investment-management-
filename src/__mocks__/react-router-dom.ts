export const useParams = jest.fn();
export const useLocation = jest.fn();
