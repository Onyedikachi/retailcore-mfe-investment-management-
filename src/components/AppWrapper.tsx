import { Provider } from "react-redux";
import { PersistGate } from "redux-persist/integration/react";
import { persistStore } from "redux-persist";
import { ToastContainer, Zoom } from "react-toastify";
import { store } from "../config/store";

function AppWrapper({ children }) {
  let persistor = persistStore(store());

  return (
    <div data-testid="app">
      <Provider data-testid="store" store={store()}>
        <PersistGate loading={<div>loading...</div>} persistor={persistor}>
          <ToastContainer
            position="top-right"
            autoClose={4000}
            draggable={false}
            // icon={false}
            transition={Zoom}
          />
          {children}
        </PersistGate>
      </Provider>
    </div>
  );
}

export default AppWrapper;
