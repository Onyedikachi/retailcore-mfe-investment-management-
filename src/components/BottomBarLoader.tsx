import React from "react";

const BottomBarLoader = () => {
  return (
    <div
      data-testid="bottom-bar-loader"
      className="w-full  p-2 flex justify-center items-center"
    >
      <div className="spinner-border h-10 w-10 border-t border-danger-500 rounded-full animate-spin"></div>
    </div>
  );
};

export default BottomBarLoader;
