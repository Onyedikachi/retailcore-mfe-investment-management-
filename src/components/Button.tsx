import { forwardRef } from "react";
import clsx from "clsx";
import { BiLoaderAlt } from "react-icons/bi";

type Ref = HTMLButtonElement;

export type ButtonProps = React.DetailedHTMLProps<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
> & {
  as?: React.ElementType;
  isLoading?: boolean;
  disabled?: boolean;
  type?: "button" | "submit" | "reset"; // Add the default value for type prop
};

const Button = forwardRef<Ref, ButtonProps>((props, ref) => {
  const {
    as: Component = "button", // Use "button" as default if as prop is not provided
    type = "button",
    className,
    children,
    isLoading,
    disabled,
    ...rest
  } = props;
  const merged = clsx("btn", className);

  return (
    <Component
      ref={ref}
      type={type}
      className={merged}
      {...rest}
      disabled={disabled}
    >
      {children}
      {isLoading && (
        <span>
          <BiLoaderAlt data-testid="loader" className="animate-spin" />
        </span>
      )}
    </Component>
  );
});

Button.displayName = "Button";
export default Button;
