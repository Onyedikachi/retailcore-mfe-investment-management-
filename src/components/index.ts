export { default as Breadcrumbs } from "./Breadcrumbs";
export { default as DropDown } from "./DropDown";
export { default as Button } from "./Button";
export { default as Loader } from "./Loader";
export { default as MultiRangeSlider } from "./MultiRangeSlider";
export { default as ProcessingStatusSlider } from "./ProcessingStatusSlider";
export { default as TopBar } from "./TopBar";
export { default as StatusCard } from "./StatusCard";
export { default as TableComponent } from "./investment/dashboard/TableComponent";
export { default as QuickLinks } from "./QuickLinks";
export { default as FormStepComponent } from "./FormStepComponent";
export { default as ProductSearch } from "./ProductSearch";
