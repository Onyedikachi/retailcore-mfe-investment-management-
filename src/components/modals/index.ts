export { default as Confirm } from "./Confirm";
export { default as Layout } from "./Layout";
export { default as ApprovedSuccess } from "./ApprovedSuccess";
export * from "./Status";
