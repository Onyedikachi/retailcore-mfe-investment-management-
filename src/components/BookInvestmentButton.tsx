import React, { useState } from "react";
import OutsideClickHandler from "react-outside-click-handler";
import { BookInvestmentOptions, IBookInvestmentOptions } from "@app/constants";
import { useNavigate } from "react-router-dom";

type BookInvestmentButton = {
  children?: any;
  disabled?: boolean;
};

export default function BookInvestmentButton({
  children,
  disabled,
}: BookInvestmentButton) {
  const navigate = useNavigate();
  const [isOpen, setIsOpen] = useState(false);

  const goToUrl = (url) => {
    if (disabled) return;
    navigate(url);

    setIsOpen(false);
  };

  return (
    <div>
      <OutsideClickHandler
        onOutsideClick={() => {
          setIsOpen(false);
        }}
      >
        <div className="relative">
          <div onKeyUp={() => {}} onClick={() => setIsOpen(true)}>{children}</div>
          {isOpen && (
            <div className="top-0 absolute w-[286px] z-10 bg-white rounded-b-lg shadow-[0px_0px_4px_0px_rgba(0,0,0,0.25)] pt-[1px]">
              <ul>
                {BookInvestmentOptions?.map((item: IBookInvestmentOptions) => (
                  <li className="group" key={item.title}>
                    <button
                      disabled={disabled}
                      onClick={() => {
                        goToUrl(item.url);
                      }}
                      className={`disabled:opacity-50 disabled:cursor-not-allowed w-full z-10 relative text-base text-[#636363] capitalize pl-6 pr-4 flex justify-between items-center py-3 group-last:rounded-b-lg hover:bg-[#F9E5E5] cursor-pointer`}
                    >
                      {item.title}
                    </button>
                  </li>
                ))}
              </ul>
            </div>
          )}
        </div>
      </OutsideClickHandler>
    </div>
  );
}
