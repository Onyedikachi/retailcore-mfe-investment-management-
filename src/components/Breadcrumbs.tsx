import React from "react";
import { BsArrowLeft } from "react-icons/bs";
import { useNavigate } from "react-router-dom";

interface BreadcrumbItem {
  id: string | number;
  title: string;
  url: string;
}

interface BreadcrumbsProps {
  links: BreadcrumbItem[];
}

export default function Breadcrumbs({
  links,
}: BreadcrumbsProps): React.ReactElement {
  const navigate = useNavigate();
  return (
    <div className="flex items-center gap-x-2" data-testid="breadcrumbs">
      <button
        className="py-1 px-1 h-8 w-8 rounded flex items-center justify-center outline-none bg-transparent hover:bg-[#F9F2F2]"
        onClick={() => navigate(-1)}
        data-testid="back button"
      >
        <BsArrowLeft className="text-[#636363]" />{" "}
      </button>
      <ul className="flex gap-x-8 uppercase items-center">
        {links.map((item: BreadcrumbItem, idx: number) => (
          <li
            key={`${idx.toString()}-index`}
            onMouseOver={() => { }}
            onClick={() => navigate(item.url)}
            data-testid={item.title}
            className={`first-of-type:hover:bg-[#F9F2F2] ${item.url !== "#" ? "hover:bg-[#F9F2F2] cursor-pointer" : ""
              } px-2 py-1 rounded first-of-type:cursor-pointer relative text-[#8F8F8F] text-base font-medium last-of-type:text-[#636363] after:absolute after:content-[''] after:right-[-16px] after:top-1/2 after:translate-y-[-50%] after:w-2 after:h-2 after:border-b-2 after:border-r-2 after:border-[#636363] after:-rotate-45 last-of-type:after:content-none`}
          >
            <span>{item.title}</span>
          </li>
        ))}
      </ul>
    </div>
  );
}
