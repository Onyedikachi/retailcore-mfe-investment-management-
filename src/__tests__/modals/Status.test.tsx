import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Success, Prompt, Failed } from "../../components/modals";
import { useNavigate, useLocation, BrowserRouter } from "react-router-dom";

jest.mock("react-router-dom", () => ({
  BrowserRouter: ({ children }) => <div>{children}</div>,
  Routes: ({ children }) => <div>{children}</div>,
  Route: ({ element }) => element,
  useNavigate: jest.fn(),
  useParams: jest.fn(),
  useLocation: jest.fn(),
}));
const navigate = jest.fn();

jest
  .spyOn(require("react-router-dom"), "useNavigate")
  .mockReturnValue(navigate);
jest.mock("react-router-dom");
class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

describe("Success Modal", () => {
  beforeAll(() => {
    window.ResizeObserver = ResizeObserver;
  });
  it("renders without crashing", async () => {
    render(
      <BrowserRouter>
        <Success
          isOpen={false}
          setIsOpen={jest.fn()}
          text={""}
        />
      </BrowserRouter>
    );
  });

  it("matches snapshot", async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Success
          isOpen={false}
          setIsOpen={function (isOpen: boolean): void {
            throw new Error("Function not implemented.");
          }}
          text={""}
        />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("can close modal pathname: /branch-management", async () => {
    jest
      .spyOn(require("react-router-dom"), "useLocation")
      .mockReturnValue({ pathname: "/branch-management" });

    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <Success isOpen={true} setIsOpen={setIsOpen} text={""} canClose />
    );
    const closebtn = await findByTestId("close-btn");
    fireEvent.click(closebtn);
    // expect(window.location.reload).toHaveBeen();
  });
  it("can close modal pathname !== /branch-management", async () => {
    jest
      .spyOn(require("react-router-dom"), "useLocation")
      .mockReturnValue({ pathname: "/" });

    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <Success isOpen={true} setIsOpen={setIsOpen} text={""} canClose />
    );
    const closebtn = await findByTestId("dashboard-link");
    fireEvent.click(closebtn);
  });
  it("can close modal when canclose is true", async () => {
    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <Success isOpen={true} setIsOpen={setIsOpen} text={"hello"} canClose />
    );
    const closebtn = await findByTestId("can-close");
    expect(closebtn).toBeInTheDocument();
    fireEvent.click(closebtn);
    expect(setIsOpen).toHaveBeenCalled();
  });
});

describe("Prompt Modal", () => {
  beforeAll(() => {
    window.ResizeObserver = ResizeObserver;
  });
  it("renders without crashing", async () => {
    render(
      <BrowserRouter>
        <Success
          isOpen={false}
          setIsOpen={function (isOpen: boolean): void {
            throw new Error("Function not implemented.");
          }}
          text={""}
        />
      </BrowserRouter>
    );
  });

  it("matches snapshot", async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Prompt
          isOpen={false}
          setIsOpen={function (isOpen: boolean): void {
            throw new Error("Function not implemented.");
          }}
          text1={""}
        />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("can close modal pathname: /branch-management", async () => {
    jest
      .spyOn(require("react-router-dom"), "useLocation")
      .mockReturnValue({ pathname: "/branch-management" });

    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <Prompt isOpen={true} setIsOpen={setIsOpen} canClose text1={""} />
    );
    const closebtn = await findByTestId("close-btn");
    fireEvent.click(closebtn);
    expect(setIsOpen).toHaveBeenCalled();
  });

  it("can close modal pathname !== /branch-management", async () => {
    jest
      .spyOn(require("react-router-dom"), "useLocation")
      .mockReturnValue({ pathname: "/" });

    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <BrowserRouter>
        {" "}
        <Prompt isOpen={true} setIsOpen={setIsOpen} canClose text1={""} />{" "}
      </BrowserRouter>
    );
    const closebtn = await findByTestId("dashboard-link");
    fireEvent.click(closebtn);
    // expect(window.location.href).toHaveBeenCalled();
  });

  it("can close modal when canclose is true", async () => {
    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <BrowserRouter>
        {" "}
        <Prompt
          isOpen={true}
          setIsOpen={setIsOpen}
          text1={"hello"}
          canClose
        />{" "}
      </BrowserRouter>
    );
    const closebtn = await findByTestId("can-close");
    expect(closebtn).toBeInTheDocument();
    fireEvent.click(closebtn);
    expect(setIsOpen).toHaveBeenCalled();
  });
});

describe("Failed Modal", () => {
  beforeAll(() => {
    window.ResizeObserver = ResizeObserver;
  });

  it("matches snapshot", async () => {
    const { asFragment } = render(
      <BrowserRouter>
        {" "}
        <Failed
          isOpen={false}
          setIsOpen={function (isOpen: boolean): void {
            throw new Error("Function not implemented.");
          }}
          text={""}
          subtext={""}
        />{" "}
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("can close modal pathname: /branch-management", async () => {
    jest
      .spyOn(require("react-router-dom"), "useLocation")
      .mockReturnValue({ pathname: "/branch-management" });

    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <BrowserRouter>
        {" "}
        <Failed
          isOpen={true}
          setIsOpen={setIsOpen}
          canClose
          text={""}
          subtext={""}
        />{" "}
      </BrowserRouter>
    );
    const closebtn = await findByTestId("close-btn");
    fireEvent.click(closebtn);
    expect(setIsOpen).not.toHaveBeenCalled();
  });


  it("can close modal pathname !== /branch-management", async () => {
    jest
      .spyOn(require("react-router-dom"), "useLocation")
      .mockReturnValue({ pathname: "/" });

    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <BrowserRouter>
        {" "}
        <Failed
          isOpen={true}
          setIsOpen={setIsOpen}
          canClose
          text={""}
          subtext={""}
        />{" "}
      </BrowserRouter>
    );
    const closebtn = await findByTestId("dashboard-link");
    fireEvent.click(closebtn);
    // expect(window.location.href).toHaveBeenCalled();
  });

  it("can close modal when canclose is true", async () => {
    const setIsOpen = jest.fn();
    const { findByTestId } = render(
      <BrowserRouter>
        {" "}
        <Failed
          isOpen={true}
          setIsOpen={setIsOpen}
          text={"hello"}
          canClose
          subtext={""}
        />{" "}
      </BrowserRouter>
    );
    const closebtn = await findByTestId("can-close");
    expect(closebtn).toBeInTheDocument();
    fireEvent.click(closebtn);
    expect(setIsOpen).toHaveBeenCalled();
  });
});
