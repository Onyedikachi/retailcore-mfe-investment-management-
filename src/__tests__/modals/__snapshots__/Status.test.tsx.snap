// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Failed Modal matches snapshot 1`] = `
<DocumentFragment>
  <div>
      
  </div>
</DocumentFragment>
`;

exports[`Prompt Modal matches snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`Success Modal matches snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
